enhancements/storage/csi-vsphere-operator.md

```diff
@@ -0,0 +1,159 @@
+
+One downside of having storage policy as an optional field is - when in-tree vSphere driver is removed from Kubernetes then without a default storage policy
+configured in OCP - such clusters will come up with no default storageClass installed and will require manual creation of storageClasses.
+
+### Deployment strategy
+
+vSphere CSI driver operator will be deployed by [cluster-storage-operator](https://github.com/openshift/cluster-storage-operator/)
+
+#### Operator deployment via cluster-storage-operator
+
+The cluster-storage-operator will deploy all the resources necessary for creating the vSphere CSI driver operator.
+
+1. vSphere CSI driver operator will be deployed in namespace `openshift-cluster-csi-drivers`.
+2. A service account will be created for running the operator.
+3. The operator will get RBAC rules necessary for running the operator and its operand (i.e the actual vSphere CSI driver).
+4. A deployment will be created for starting the operator and will be managed by cluster-storage-operator.
+5. A instance of `ClusterCSIDriver` will be created to faciliate managment of driver operator.  `ClusterCSIDriver` is already defined in - https://github.com/openshift/api/blob/master/operator/v1/types_csi_cluster_driver.go but needs to be expanded to include vSphere CSI driver.
+6. cluster-storage-operator will create required cloud-credentials for talking with vCenter API.
+
+#### Driver deployment via vsphere-csi-driver-operator.
+
+The operator itself will be responsible for running the driver and all the required sidecars (attacher, provisioner etc).
+
+1. The operator will create namespace `openshift-cluster-csi-drivers` to deploy the driver and sidecars.
+2. A service account will be created for running the driver.
+3. The operator will create RBAC rules necessary for running the driver and sidecars.
+4. A deployment will be created and managed by operator to handle control-plane sidecars and controller-plane driver deployment with controller-side of CSI services.
+5. A DaemonSet will be created and managed by operator to run node side of driver operations.
+6. A `CSIDriver` object will be created to expose driver features to control-plane.
+7. The driver operator will use and expose cloud-credentials created by cluster-storage-operator.
+
+Most of the steps outlined above is common to all CSI driver operator and vSphere CSI driver is not unique among those aspects.
+
+#### Additional consideration for vSphere
+
+There are certain aspects of driver which require special handling:
+
+1. When vSphere CSI operator starts, using credentials provided by cluster-storage-operator, it will first verifiy vCenter version and HW version of VMs. If vCenter version is not 6.7u3 or greater and HW version is not 15 or greater on all VMs - it will set `vSphereOperatorDisabled: true` and stop further processing.
```

Comment:
What if a new VM with HW ver < 15 is added later? Should it un-install the existing driver too? I would expect that the driver becomes `Available=False`, maybe even `Degraded=True` with an alert or something, but the operator should keep resyncing periodically.

It could be even useful to label nodes that are able to run the driver and run the driver only there.

Reply:
Yeah that is a good point. I think we will want to mark operator as Degraded but it would still keep resyncing and existing pods/volumes should keep working. But it also means that, if a pod wants to be scheduled to a node which is on HW<15, then that node can't run the pod.

Reply:
added some language about degraded driver status when this happens.